db/blob/blob_fetcher.h

```diff
@@ -32,6 +33,16 @@ class BlobFetcher {
 
  private:
   const Version* version_;
-  ReadOptions read_options_;
+  const ReadOptions& read_options_;
 };
+
+class BlobFetcherWithReadOptionsCopy : public BlobFetcher {
+ public:
+  BlobFetcherWithReadOptionsCopy(const Version* v, const ReadOptions& ro)
+      : BlobFetcher(v, read_options_copy_), read_options_copy_(ro) {}
```

Comment:
If `BlobFetcher` uses this reference in its constructor, it will cause some hard to find bugs. This subclass is enforcing an unusual requirement on its base class.  Future users of `BlobFetcher` needs to be conscious of the scope of their `ReadOptions` relative to their `BlobFetcher` in order to choose the right one of the two. This at least requires more documentation of this class.  I wonder if the copying saved is worth introducing this pattern. Checking with @ltamasi to make a call here.

Reply:
Thanks @jowlyzhang ! I agree that reference members are pretty accident prone: as an example, if `BlobFetcher` ever started to use `read_options_` in its destructor, `BlobFetcherWithReadOptionsCopy` would break since the `ReadOptions` in the derived class would have been destroyed by the time the base class destructor runs.

@rockeet The goal of the PR is to save the `ReadOptions` copy that happens in `Version::Get` and `Version::MultiGet`, right? We could achieve that by changing `GetContext` so it takes a back pointer to `Version` (instead of a `BlobFetcher` pointer) and calls `Version::GetBlob` directly.

Reply:
> We could achieve that by changing `GetContext` so it takes a back pointer to `Version` (instead of a `BlobFetcher` pointer) and calls `Version::GetBlob` directly.

Hm, actually, scratch that - the key issue is how we're going to access the `ReadOptions` struct since `GetContext` doesn't have access to it either. Yeah, saving this copy might not be worth the extra complexity then.